ragna/ui/resources/upload.js

```diff
@@ -1,18 +1,25 @@
 
 
 function upload(files, token, informationsEndpoint, final_callback) {
+
     var uploaded_documents = [] 
 
     Array.from(files).map(file => {
 
         getUploadInformations(file, token, informationsEndpoint).then(
 
                 function(uploadInformations) {
-                        uploadFile(uploadInformations.url, uploadInformations.data, file)
-                        uploaded_documents.push(uploadInformations.document)
-                        if (uploaded_documents.length == files.length) {
-                            final_callback(uploaded_documents);
-                        }
+
+                        uploadFile(uploadInformations.url, uploadInformations.data, file).then(
+                            function(response) {
```

Comment:
@pierrotsmnrd While your patch here fixes the waiting for the upload to be completed, it reverts one that I made earlier (see https://github.com/Quansight/ragna/pull/61#discussion_r1367991066). To reiterate: we **cannot** use the `response` of the upload to fill the `uploaded_documents`. Our local upload endpoint indeed returns the document data, i.e. id and name. This response is "user configured".

We need to revert back to what I had, but need to wait for the upload to complete. Maybe it is as easy as `uploadFile(...).then(function(response) {})`?

Reply:
@pierrotsmnrd I've ventured into JS again and refactored the whole file. With the snippet below we now perform the upload in parallel since this is heavily I/O bound.

```js
function upload(files, token, informationsEndpoint, final_callback) {
  Promise.all(
    Array.from(files).map((file) => {
      return uploadFile(file, token, informationsEndpoint);
    }),
  ).then(final_callback);
}

async function uploadFile(file, token, informationEndpoint) {
  const response = await fetch(`${informationEndpoint}?name=${file.name}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  const documentInfo = await response.json();

  var body = new FormData();
  for (const [key, value] of Object.entries(documentInfo.data)) {
    body.append(key, value);
  }
  body.append("file", file);

  await fetch(documentInfo.url, {
    method: "POST",
    body: body,
  });

  return documentInfo.document;
}
```

The signature of the `upload` function is the same, so there is no need to change anything besides this file. Please have a look and LMK if we can go for this. I made sure we properly await everything this time :innocent:

Reply:
Looks good to me, tested and approved :) I'm going to commit and credit you